torch_xla/core/xla_builder.py

```diff
@@ -855,32 +860,141 @@ def jax_func_to_xla_computation(jax_func, args, kwargs, name=None):
+      import torch_xla
+      if a is None:
+        return None
+      if isinstance(a, torch.Tensor):
+        assert a.device == torch_xla.device(
+        ), f"Inputs must be XLA tensors. Got {a.device}"
+        return jax.ShapeDtypeStruct(a.shape, mappings.t2j_dtype(a.dtype))
+      return a
+
+    sample_inputs = tuple(abstractify(a) for a in flattened_inputs)
+
+    # Pick out the non-static args.
+    # Consider anything that is not a `jax.ShapeDtypeStruct` as a static arg.
+    def pick_tensor_args(flattened_args):
+      tensor_args = []
+      for i in range(len(sample_inputs)):
+        if isinstance(sample_inputs[i], jax.ShapeDtypeStruct):
+          tensor_args.append(flattened_args[i])
+      return tensor_args
+
+    sample_tensor_args = pick_tensor_args(sample_inputs)
+    tensor_args = pick_tensor_args(flattened_inputs)
+
+    # This function only takes in tensor arguments because its signature must
+    # match the signature of the HLO module lowered from JAX, allowing us to
+    # wrap it in an XLA user computation.
+    def fn(*tensor_args):
+      # Go from a list of tensor args to the full list of flattened arguments,
+      # by referencing the original flattened inputs.
+      new_flattened = copy(flattened_inputs)
+      tensor_args_iter = iter(tensor_args)
+      for i in range(len(sample_inputs)):
+        if isinstance(sample_inputs[i], jax.ShapeDtypeStruct):
+          new_flattened[i] = next(tensor_args_iter)
+      args, kwargs = jax.tree.unflatten(spec, new_flattened)
+      return jax_func(*args, **kwargs)
+
+    def get_hlo():
+      import torch_xla.debug.profiler as xp
+      # If we see this trace span in the profiler, we'll know that there's a cache miss.
+      with xp.Trace('jax_to_hlo'):
+        hlo_ir = jax.jit(
+            fn, keep_unused=True).lower(*sample_tensor_args).compiler_ir('hlo')
+
+        # Get a protobuf representation of the HLO. `as_serialized_hlo_module_proto` is
+        # mentioned at https://github.com/jax-ml/jax/discussions/22266
+        return hlo_ir.as_serialized_hlo_module_proto()  # type: ignore
+
+    hlo_module = _jax_to_hlo_cache_get_or_insert(jax_func, sample_inputs, spec,
+                                                 get_hlo)
+    return XlaComputation(name, hlo_module, tensor_args, pick_tensor_args)
+
+
+def _jax_to_hlo_cache_get_or_insert(jax_func, sample_inputs: tuple[Any, ...],
+                                    input_tree_spec, get_hlo):
+  # Use two layers of dictionary lookup.
+  # The first layer uses the `jax_func`, which is only weakly referenced.
+  # The second layer uses the sample inputs and the tree spec, which is strongly referenced.
+  inner_dict = _JAX_TO_HLO_CACHE.get(jax_func, None)
+  if inner_dict is not None:
+    hlo = inner_dict.get((sample_inputs, input_tree_spec), None)
+    if hlo is not None:
+      return hlo
+
+  # Compget_hlo jax function to HLO.
+  hlo = get_hlo()
+  if inner_dict is None:
+    _JAX_TO_HLO_CACHE[jax_func] = {}
```

Comment:
you might need to declare `global _JAX_TO_HLO_CACHE` here

Reply:
done